src/v/kafka/server/handlers/sasl_handshake.cc

```diff
@@ -32,23 +62,22 @@ ss::future<response_ptr> sasl_handshake_handler::handle(
      * authenticate request for the next phase of the auth process.
      */
     auto error = error_code::none;
-
-    if (request.data.mechanism == security::scram_sha256_authenticator::name) {
-        ctx.sasl()->set_mechanism(
-          std::make_unique<security::scram_sha256_authenticator::auth>(
-            ctx.credentials()));
-
-    } else if (
-      request.data.mechanism == security::scram_sha512_authenticator::name) {
-        ctx.sasl()->set_mechanism(
-          std::make_unique<security::scram_sha512_authenticator::auth>(
-            ctx.credentials()));
-
+    auto mechs = supported_mechanisms();
+    auto const& it = mechs.find(request.data.mechanism);
+    if (it != mechs.end()) {
+        ctx.sasl()->set_mechanism(it->second(ctx.credentials()));
```

Comment:
does this build instances of all the mechanisms but then only choose one of them? if so, why not a few if/else statements instead?

Reply:
It builds a map of factories.

Reply:
I've simplified this a bit.